src/serialize.rs

```diff
@@ -180,3 +210,185 @@ pub fn compute_serialized_size_in_words<U : MessageBuilder>(message: &mut U) ->
+
+        let mut buf = vec![];
+
+        buf.extend([0,2,0,0].iter().cloned()); // 513 segments
+        buf.extend([0; 513 * 4].iter().cloned());
+        assert!(read_frame(&mut Cursor::new(&buf[..]),
+                           ReaderOptions::new()).is_err());
+        buf.clear();
+
+        buf.extend([0,0,0,0].iter().cloned()); // 1 segments
+        assert!(read_frame(&mut Cursor::new(&buf[..]),
+                           ReaderOptions::new()).is_err());
+        buf.clear();
+
+        buf.extend([0,0,0,0].iter().cloned()); // 1 segments
+        buf.extend([0; 3].iter().cloned());
+        assert!(read_frame(&mut Cursor::new(&buf[..]),
+                           ReaderOptions::new()).is_err());
+        buf.clear();
+
+        buf.extend([255,255,255,255].iter().cloned()); // 0 segments
+        assert!(read_frame(&mut Cursor::new(&buf[..]),
+                           ReaderOptions::new()).is_err());
+        buf.clear();
+    }
+
+    #[test]
+    fn test_write_frame() {
+
+        let mut buf = vec![];
+
+        let segment_0 = [Word::from(0); 0];
+        let segment_1 = [Word::from(1); 1];
+        let segment_199 = [Word::from(199); 199];
+
+        write_frame(&mut buf, &[&segment_0]).unwrap();
+        assert_eq!(&[0,0,0,0,  // 1 segments
+                     0,0,0,0], // 0 length
+                   &buf[..]);
+        buf.clear();
+
+        write_frame(&mut buf, &[&segment_1]).unwrap();
+        assert_eq!(&[0,0,0,0,  // 1 segments
+                     1,0,0,0], // 1 length
+                   &buf[..]);
+        buf.clear();
+
+        write_frame(&mut buf, &[&segment_199]).unwrap();
+        assert_eq!(&[0,0,0,0,    // 1 segments
+                     199,0,0,0], // 199 length
+                   &buf[..]);
+        buf.clear();
+
+        write_frame(&mut buf, &[&segment_0, &segment_1]).unwrap();
+        assert_eq!(&[1,0,0,0,  // 2 segments
+                     0,0,0,0,  // 0 length
+                     1,0,0,0,  // 1 length
+                     0,0,0,0], // padding
+                   &buf[..]);
+        buf.clear();
+
+        write_frame(&mut buf, &[&segment_199, &segment_1, &segment_199, &segment_0]).unwrap();
+        assert_eq!(&[3,0,0,0,   // 4 segments
+                     199,0,0,0, // 199 length
+                     1,0,0,0,   // 1 length
+                     199,0,0,0, // 199 length
+                     0,0,0,0,   // 0 length
+                     0,0,0,0],  // padding
+                   &buf[..]);
+        buf.clear();
+
+        write_frame(&mut buf, &[&segment_199, &segment_1, &segment_199, &segment_0, &segment_1]).unwrap();
+        assert_eq!(&[4,0,0,0,   // 5 segments
+                     199,0,0,0, // 199 length
+                     1,0,0,0,   // 1 length
+                     199,0,0,0, // 199 length
+                     0,0,0,0,   // 0 length
+                     1,0,0,0],  // 1 length
+                   &buf[..]);
+        buf.clear();
+    }
+
+    #[test]
+    fn check_serde() {
```

Comment:
When I first read this code I was confused for a bit about what it had to do with [serde](https://github.com/serde-rs/serde). I think "round_trip" might be a more clear name.